third_party/move/move-model/src/ty.rs

```diff
@@ -1086,6 +1086,77 @@ impl Type {
         matches!(self, Type::Var(_))
     }
 
+    /// Returns all internal types contained in this type (including itself), skipping reference types.
+    pub fn get_all_contained_types_with_skip_reference(&self, env: &GlobalEnv) -> Vec<Type> {
+        match self {
+            Type::Primitive(p) => vec![Type::Primitive(*p)],
+            Type::Tuple(ts) => ts
+                .iter()
+                .flat_map(|t| t.get_all_contained_types_with_skip_reference(env))
+                .collect(),
+            Type::Vector(et) => {
+                let mut types = et.get_all_contained_types_with_skip_reference(env);
+                types.push(self.clone());
+                types
+            },
+            Type::Struct(_, _, ts) => {
+                let mut new_types = ts
```

Comment:
Maybe mention that phantom types are also included?

Reply:
good catch! modified the code such that phantom types are not included.